File: python/06_valid_anagram.py
"""
Q6. Valid Anagram (Category: Strings)

Problem: Return True if t is an anagram of s.

Approach: Frequency count array of size 26 (lowercase letters only).
          Increment for s, decrement for t. All must be zero at the end.

Time Complexity : O(n)
Space Complexity: O(1)
"""


def is_anagram(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False
    count = [0] * 26
    for a, b in zip(s, t):
        count[ord(a) - ord("a")] += 1
        count[ord(b) - ord("a")] -= 1
    return all(c == 0 for c in count)


if __name__ == "__main__":
    print(is_anagram("anagram", "nagaram"))  # True
    print(is_anagram("rat", "car"))          # False

File: cpp/06_ValidAnagram.cpp
/*
 * Q6. Valid Anagram (Category: Strings)
 *
 * Problem: Return true if t is an anagram of s.
 *
 * Approach: Frequency count array of size 26.
 *           Increment for s, decrement for t. All must be zero.
 *
 * Time Complexity : O(n)
 * Space Complexity: O(1)
 */
#include <iostream>
#include <string>
#include <vector>
using namespace std;

bool isAnagram(string s, string t) {
    if (s.length() != t.length()) return false;

    vector<int> count(26, 0);
    for (int i = 0; i < (int)s.length(); i++) {
        count[s[i] - 'a']++;
        count[t[i] - 'a']--;
    }
    for (int c : count) {
        if (c != 0) return false;
    }
    return true;
}

int main() {
    cout << boolalpha;
    cout << isAnagram("anagram", "nagaram") << endl; // true
    cout << isAnagram("rat", "car")         << endl; // false
    return 0;
}
